software/spmd/remote/main.cpp

```diff
@@ -0,0 +1,181 @@
+float C[N_ELS] __attribute__ ((section (".dram"))) = {0};
+float gold[N_ELS] __attribute__ ((section (".dram")));
+
+float bsg_attr_remote _A[N_ELS] __attribute__ ((section (".dram"))) = MACRO_A;
+float bsg_attr_remote _B[N_ELS] __attribute__ ((section (".dram"))) = MACRO_B;
+float bsg_attr_remote _C[N_ELS] __attribute__ ((section (".dram"))) = {0};
+float bsg_attr_remote _gold[N_ELS] __attribute__ ((section (".dram"))) = {0};
+
+int main(){
+
+        bsg_set_tile_x_y();
+
+        if ((__bsg_x == 0) && (__bsg_y == 0)) {
+                saxpy(A, B, gold, MACRO_ALPHA); 
+ 
+                // **************************************************
+                // Test the simple version of SAXPY
+                // **************************************************
+                bsg_printf("Testing SAXPY\n");
+                // Heat the cache (just for demonstration purposes)
+                saxpy(A, B, C, MACRO_ALPHA);
+                // Measure performance
+                bsg_cuda_print_stat_start(SAXPY_TAG_BASE);
+                saxpy(A, B, C, MACRO_ALPHA);
+                bsg_cuda_print_stat_end(SAXPY_TAG_BASE);
+                
+                if (memcmp(C, gold, N_ELS * sizeof(*C)))
+                        bsg_fail();
+                memset(C, 0, N_ELS * sizeof(*C));
+
+                // **************************************************
+                // Test the C version of SAXPY
+                // **************************************************
+                bsg_printf("Testing C SAXPY +unroll\n");
+                // Heat the cache (just for demonstration purposes)
+                saxpy_c(A, B, C, MACRO_ALPHA);
+                // Measure performance
+                bsg_cuda_print_stat_start(SAXPY_TAG_C_UNROLL);
+                saxpy_c(A, B, C, MACRO_ALPHA);
+                bsg_cuda_print_stat_end(SAXPY_TAG_C_UNROLL);
+                
+                if (memcmp(C, gold, N_ELS * sizeof(*C)))
+                        bsg_fail();
+                memset(C, 0, N_ELS * sizeof(*C));
+
+
+                // **************************************************
+                // Test the C version of SAXPY, with remote
+                // **************************************************
+                bsg_printf("Testing C SAXPY +unroll +remote\n");
+                // First generate gold values
+                saxpy_c_remote(_A, _B, _gold, MACRO_ALPHA);
+
+                // Heat the cache (just for demonstration purposes)
+                saxpy_c_remote(_A, _B, _C, MACRO_ALPHA);
+                // Measure performance
+                bsg_cuda_print_stat_start(SAXPY_TAG_C_UNROLL_REMOTE);
+                saxpy_c_remote(_A, _B, _C, MACRO_ALPHA);
+                bsg_cuda_print_stat_end(SAXPY_TAG_C_UNROLL_REMOTE);
+
+#ifndef __llvm__
+                // LLVM complains about memory intrinsics...
+                if (memcmp(_C, _gold, N_ELS * sizeof(*C)))
```

Comment:
There's a pragma to turn off intrinsics with llvm if it's helpful here...

Reply:
It would be!

Reply:
There's a flag actually `-fno-builtin` to force llvm to call actual memcpy instead of inserting an intrinsic. I vaguely remember that there are pragmas to disable specific intrinsic too.

Reply:
https://clang.llvm.org/docs/AttributeReference.html#no-builtin

Reply:
This still didn't fix the issue (I think): 
`/mnt/users/spin0/no_backup/drichmond/Research/repositories/git/bsg_bladerunner/bsg_manycore/software/riscv-tools/riscv-install/bin/../riscv32-unknown-elf-dramfs/include/string.h:30:7: note: candidate function not viable: no known conversion from 'bsg_attr_remote float [512]' to 'const void *' for 1st argument
int      memcmp (const void *, const void *, size_t);`